### PR TITLE
Template Parts: allow selection in page content focus mode

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Allows Template Part blocks (Header and Footer) to be selected when editing a page in the site editor in page content focus mode.

## Why?
See https://github.com/WordPress/gutenberg/issues/59970#issuecomment-2005483531.

## How?
- Set the editing mode of `core/template-part` blocks to `contentOnly`, so that you can select it.
- Set the editing mode of children of `core/template-part` blocks to `disabled`, so that you're forced to click _Edit_ in the toolbar.
- Changes `BlockToolbar` to show the block icon / block name on blocks that have a `contentOnly` editing mode, so that you can see the name of the template part when you've selected it.

## Testing Instructions
1. Go to Appearance → Editor → Pages and create or edit page.
2. Click on the header or footer.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/612155/6d9f9c36-e45b-4f6f-befc-5869b8bb2976

### CODE REVIEW

**Review by SaxonF — COMMENTED** [review]
Working well in testing. Nice work! I think its an improvement and finds a better balance whilst still locking templates which aren't modified as much as site navigation. @richtabor you happy to start here?

File: packages/block-editor/src/components/block-switcher/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/block-switcher/index.js]: Tiny nit: Is it worth adding a simple comment before this line? Previously the `hideDropdown` const communicated that this block was about outputting the components without the dropdown, so it might help readability slightly to have a comment in its place?

**Review by andrewserong — COMMENTED** [review]
Nice, this feels pretty good to use to me @noisysocks, so feels like a good place to start IMO! A couple of minor details:

The test `create a new page, edit template and toggle page template preview ` appears to be failing expecting the `Edit your template to edit this block` notice, so some of the e2es might need updating.

While this is testing nicely for me in the site editor's view of editing a page, in the post editor with template preview mode switched on, it's possible to select the header or footer, and they're visible in the list view, but there's no option to Edit, so it isn't really possible to do anything. In this case, if the blocks cannot be edited in the current context, what would the desired behaviour be?

https://github.com/WordPress/gutenberg/assets/14988353/c3022c79-94f6-496d-8603-7efb6e8bcc61

**Comment by jasmussen** [comment]
Nice work. When you're editing a page, this seems like a good flow. To clarify a hidden nuance here, which is related to editing of inline template parts as [also discussed here](https://github.com/WordPress/gutenberg/pull/59883#issuecomment-2005006973): in this branch inline editing of template parts is still possible when you select a _template_ in the site editor. So the main change here is for _pages only_, where now you can select the TP, but only edit it in isolation.

Since we are already having a special case for pages, this seems okay to me, but would love also a gut-check from @richtabor.

**Comment by jameskoster** [comment]
Nice, this is working well for me and significantly shortens the journey to go and edit template parts from a page 🚀  

A couple of small quirks we'll likely want to address:

* Hover / focus outline should be purple rather than theme-color. 
* When the page loads, template parts show a dotted outline which fades away – I assume due to `contentOnly`? I'm not sure we need to specifically draw attention to the edit-ability of template parts?
* The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

Screenshot 2024-03-20 at 10 00 50

* This only happens with my footer for some reason (stock `page.html` template in TT4), but when I select then focus away there's a double-border effect, one purple and one blue:

Screenshot 2024-03-20 at 10 13 38

**Comment by richtabor** [comment]
> So the main change here is for pages only, where now you can select the TP, but only edit it in isolation.

That's my understanding as well, and on point with my expectations of this effort. 

> Hover / focus outline should be purple rather than theme-color.
> When the page loads, template parts show a dotted outline which fades away – I assume due to contentOnly? I'm not sure > we need to specifically draw attention to the edit-ability of template parts?

I think the whole dotted lines fading in work needs revisiting. It's not quite landing right.

**Review by richtabor — APPROVED** [review]
This is certainly a step forward. 

I'd like to see further explorations on an inline focus/isolated view, rather than navigating to the disconnected view next.

**Comment by noisysocks** [comment]
> The test create a new page, edit template and toggle page template preview  appears to be failing expecting the Edit your template to edit this block notice, so some of the e2es might need updating.

Fixed 👍 

> While this is testing nicely for me in the site editor's view of editing a page, in the post editor with template preview mode switched on, it's possible to select the header or footer, and they're visible in the list view, but there's no option to Edit, so it isn't really possible to do anything. In this case, if the blocks cannot be edited in the current context, what would the desired behaviour be?

Great catch @andrewserong. My read on the fix for this is that we should be reducing the number of differences between the site editor and the post editor. I know @youknowriad has been doing a lot in this area. So with that in mind in 34e9348a554a190f4ce888003bd9a711fd7a3c29 I moved the code that adds the _Edit_ button to the Template Part toolbar from `@wordpress/edit-site` to `@wordpress/editor` so that it appears in both editors.

**Comment by ramonjd** [comment]
Nice stuff.

I tested this in the site and post editors. Works well with theme template parts and custom template parts.

Highlight color appears, for me, correctly on mouseover and on touch. I couldn't break it after the latest updates.

https://github.com/WordPress/gutenberg/assets/6458278/e314ddfb-1418-472e-a315-63a5f2400ebf

🚢

**Comment by andrewserong** [comment]
> So with that in mind in https://github.com/WordPress/gutenberg/commit/34e9348a554a190f4ce888003bd9a711fd7a3c29 I moved the code that adds the Edit button to the Template Part toolbar from @wordpress/edit-site to @wordpress/editor so that it appears in both editors.

Thanks for following up. Unfortunately, it doesn't appear to be working properly for me in the post editor. When I go to click Edit, the template part appears to be in content only mode. Then if I click to switch to edit the template, I get a blank screen:

https://github.com/WordPress/gutenberg/assets/14988353/e8088946-4153-40e7-b602-6c2999ef7661

Not sure what's going on there, but if it's too tricky to get working, would it work to only expose the template parts in the site editor? Or does that introduce more complexity again 🤔

**Comment by ramonjd** [comment]
> When I go to click Edit, the template part appears to be in content only mode. Then if I click to switch to edit the template, I get a blank screen

I'm noticing this too, but also in trunk

https://github.com/WordPress/gutenberg/assets/6458278/4b484c0d-88df-47e9-a3d3-97baa1db4ab4

**Comment by noisysocks** [comment]
Technically it's an existing issue (you also run into it if you select _Template preview_ and then _Edit template_) but I'll see if it's easy to include a fix here.

**edit:** done.

**Comment by noisysocks** [comment]
> Hover / focus outline should be purple rather than theme-color.
> This only happens with my footer for some reason (stock page.html template in TT4), but when I select then focus away there's a double-border effect, one purple and one blue:

I'm not seeing these issues. Maybe I'm missing something?

> The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

What's a TP sorry? 😄 **edit:** Template Part! 🤦‍♂️ Good catch, I'll look into it.

> When the page loads, template parts show a dotted outline which fades away – I assume due to contentOnly? I'm not sure we need to specifically draw attention to the edit-ability of template parts?

Yeah that's right. I agree it's not worth highlighting the template part. I struggled to come up with a good solution for this because the flashing needs to be implemented in an agnostic way that works across all locked blocks (e.g. patterns with `templateLock=contentOnly`). I'll dig a little more.

**Comment by andrewserong** [comment]
> but I'll see if it's easy to include a fix here. edit: done.

That's testing nicely for me, now! 🚀

File: packages/edit-post/src/editor.js
Comment (youknowriad) [inline packages/edit-post/src/editor.js]: This mode has been removed in https://github.com/WordPress/gutenberg/pull/58935 are we restoring it back? Why?

File: packages/editor/src/components/provider/disable-non-page-content-blocks.js
Comment (youknowriad) [inline packages/editor/src/components/provider/disable-non-page-content-blocks.js:19]: Not specific to this PR but I see this component as a hack. Also the `setBlockEditingMode` and `useBlockEditingMode` APIs have proven to be problematic in a lot of ways. Multiple conflicts and race conditions, I'd love to absorb these in the block editor framework/package instead (maybe through a setting or something) but we'd need to think properly about the API that we'd want to provide there.

File: packages/editor/src/hooks/template-part-edit.js
Comment (youknowriad) [inline packages/editor/src/hooks/template-part-edit.js]: What's (which dependency) preventing us from bundling this code within the template part block code directly?

**Comment by jameskoster** [comment]
> I'm not seeing these issues. Maybe I'm missing something?

At 5s in the video in the OP you can see the outline that appears when you hover the template part is blue, but when you select it it becomes purple. I think it should be purple on hover too. 

Similarly the outline is blue when the template part is selected but not focussed. Basically the outlines should always be purple.

**Comment by bacoords** [comment]
I just want to add that conceptually I'm a fan of this approach. I did not engage in extensive, testing, just brief. Small issues/questions I noted:

- Is there a way to still respect root padding when editing a root-level template part like header/footer? I assume there's a technical challenge there.
- Is it possible to include the "double click" to edit a template part that (I believe) is currently available?
- How can we edit the post title on pages in the site editor if the template doesn't include a post title - is putting a title block in the content the best approach? How does this effect pages where the title is not actually visible on the page content?

File: packages/edit-post/src/editor.js
Comment (noisysocks) [reply packages/edit-post/src/editor.js]: Aha my mistake. I want `post-only`. The tricky part is setting or resetting the editing mode when you navigate forward or backwards.

File: packages/editor/src/components/provider/disable-non-page-content-blocks.js
Comment (noisysocks) [reply packages/editor/src/components/provider/disable-non-page-content-blocks.js:19]: Yeah agreed and I keep running into limitations too. For example in this PR I want to disable insertion but not selection on the Template Part and disable all of its children. `contentOnly` doesn't really cut it for the former. There were similar issues with synced patterns.

We also have lots of ways to do similar things. I feel that with the right abstraction we could unify `tempateLock`, the `lock` attribute, and `setBlockEditingMode`.

File: packages/editor/src/hooks/template-part-edit.js
Comment (noisysocks) [reply packages/editor/src/hooks/template-part-edit.js]: Nothing stands out to me. I'll try it.

File: packages/edit-post/src/editor.js
Comment (noisysocks) [reply packages/edit-post/src/editor.js]: Fixed in 120c78555b8560bba67d3db5349e10a39916233f.

Also updated some docs missed in https://github.com/WordPress/gutenberg/pull/58935: c54239624a160cb4881012881775cacef095bbb3.

File: packages/editor/src/hooks/template-part-edit.js
Comment (noisysocks) [reply packages/editor/src/hooks/template-part-edit.js]: 👍 27962d05c5

**Comment by noisysocks** [comment]
> * The appending inserter shows up when you select a TP, which allows partial editing outside of focus mode:

I fixed this in 19516d7588 by disabling appending when a template part is set to `contentOnly` which is not a great solution. I think we need to iterate on `setBlockEditingMode` to accomodate cases like this, but this will do for now.

**Comment by noisysocks** [comment]
> When the page loads, template parts show a dotted outline which fades away – I assume due to contentOnly? I'm not sure we need to specifically draw attention to the edit-ability of template parts?
> Similarly the outline is blue when the template part is selected but not focussed. Basically the outlines should always be purple.

OK think I've fixed these in https://github.com/WordPress/gutenberg/pull/60010/commits/b0b35daecc133e56e6c9381f253406271800a824. Thanks @jameskoster. Feel free to push more CSS tweaks if I missed something.

**Comment by noisysocks** [comment]
> * Is there a way to still respect root padding when editing a root-level template part like header/footer? I assume there's a technical challenge there.

What do you mean by root padding?

> * Is it possible to include the "double click" to edit a template part that (I believe) is currently available?

I don't think we currently have a double click shortcut for template parts? Or at least locally in `trunk` when I double click a template part nothing happens beyond the inner block that I clicked on being selected which I'd expect to happen in any block.

> * How can we edit the post title on pages in the site editor if the template doesn't include a post title - is putting a title block in the content the best approach? How does this effect pages where the title is not actually visible on the page content?

Good question. Right now one would have to switch to the post editor if there's no Post Title block. It could be worth exploring allowing the title to be set in one of the sidebars (black W sidebar on left, or page sidebar on right). It's beyond the scope of this PR though.

**Comment by noisysocks** [comment]
Thanks everyone! I've addressed / responded to all feedback. I'll merge once there's an approving code review.

**Comment by bacoords** [comment]
> What do you mean by root padding?

I'll open a separate issue for this.

> I don't think we currently have a double click shortcut for template parts? Or at least locally in trunk when I double click a template part nothing happens beyond the inner block that I clicked on being selected which I'd expect to happen in any block.

Currently if you click the surrounding template or a template part from a page in the site editor you'll see a toast notification, but on a double click you get the modal that allows you to enter: 

Screenshot 2024-03-21 at 8 29 06 PM

The caveat being (as I think you implied) that it allows you to edit the "template" whether you've double clicked over a template or a template part. But in trunk once in the template you are at that point able to edit either one in the same context. 

In this PR you can see the same thing happen when you double click to edit the template around the page content, but not when you double click to edit the template part. That region previously responded to a double click but doesn't any longer. 

Just noting that the behavior is inconsistent and it would be great to be able to say definitely to our users "you can double click to enter into editing global content". Understand if that requires a new issue/PR.

File: packages/edit-post/src/hooks/use-navigate-to-entity-record.js
Comment (andrewserong) [inline packages/edit-post/src/hooks/use-navigate-to-entity-record.js:45]: Why is `initialPost` removed from here and moved further down into the `Editor` component? (I'm just curious, this is not a request to make any changes here 🙂)

**Review by andrewserong — APPROVED** [review]
Just gave this a re-test and it's still all working nicely for me in the post and site editors 👍 

The code change is also looking good to me, and seems consistent with the feedback that Riad gave in reviews, too, with relevant code moved over to the template part block where possible.

LGTM! ✨

File: packages/edit-post/src/hooks/use-navigate-to-entity-record.js
Comment (noisysocks) [reply packages/edit-post/src/hooks/use-navigate-to-entity-record.js:45]: It doesn't have anything to do with the primary purpose of the hook (building `onNavigateToEntityRecord` callbacks) so I split it out. Feels weird to me that you'd pass `initialPostType` and `initialPostId` to a hook and it returns exactly what you gave it in an object without doing anything else.

Comment (andrewserong) [reply packages/edit-post/src/hooks/use-navigate-to-entity-record.js:45]: Ah, gotcha. That makes sense, thanks!

**Comment by noisysocks** [comment]
> > What do you mean by root padding?
> 
> I'll open a separate issue for this.

Thank you! That'd be great.

> In this PR you can see the same thing happen when you double click to edit the template around the page content, but not when you double click to edit the template part. That region previously responded to a double click but doesn't any longer.

Ah gotcha. I'm not sure. Personally I don't really like that dialog at all 😅 I think it would be unexpected if double clicking a template part behaved differently to double clicking a different block. Maybe the designers in this thread can weigh in on what we should do here. I can address it in a follow up if need be.

**Comment by Mamaduka** [comment]
@noisysocks, one of the e2e tests became flaky after this got merged into the trunk - https://github.com/WordPress/gutenberg/issues/60102.

**Update:** Here's my attempt to fix it - #60109.

File: packages/editor/src/components/provider/disable-non-page-content-blocks.js
Comment (youknowriad) [reply packages/editor/src/components/provider/disable-non-page-content-blocks.js:19]: > We also have lots of ways to do similar things. I feel that with the right abstraction we could unify tempateLock, the lock attribute, and setBlockEditingMode.

I prefer templateLock instead because it's declarative while setBlockEditingMode is more subject to race conditions. I think the framework should be able to "guess" the mode, rather the consumer providing it for each block.

**Comment by jameskoster** [comment]
> How can we edit the post title on pages in the site editor if the template doesn't include a post title

https://github.com/WordPress/gutenberg/issues/52763

**Comment by youknowriad** [comment]
> How can we edit the post title on pages in the site editor if the template doesn't include a post title

You can also disable "template previewing" from the "template panel" in the sidebar and you'll be thrown into the regular "post only" mode that does have a "title" input (like the post editor)

File: packages/block-editor/src/components/block-toolbar/index.js
Comment (retrofox) [inline packages/block-editor/src/components/block-toolbar/index.js:174]: Removing this condition makes the block toolbar controls visible regardless of the editing mode. It affects part of the Product Editor UI. [Here is the issue](https://github.com/woocommerce/woocommerce/issues/45908).

Before this change, when the editing mode was contentOnly, the controls that didn't allow editing the block content were not visible.
Below, the custom TextArea field block shows the alignment and regular format buttons.

image

After this change, the toolbar shows the parent block, which doesn't do anything since its template is locked to contentOnly:

Screenshot 2024-04-01 at 16 43 25

Could we consider it a breaking change in terms of the UI?
How could we hide it, recreating the previous state before merging these changes?

Comment (noisysocks) [reply packages/block-editor/src/components/block-toolbar/index.js:174]: Thanks for flagging. I'll open a fix.

Comment (noisysocks) [reply packages/block-editor/src/components/block-toolbar/index.js:174]: I've opened https://github.com/WordPress/gutenberg/pull/60647 which dials this change back to only show the icon on _pattern_ blocks with `contentOnly`.

Comment (retrofox) [reply packages/block-editor/src/components/block-toolbar/index.js:174]: Nice! Thanks a bunch. I see it's already 🍏

**Pull request merged**
